Library/OcCpuLib/FrequencyDetect.c

```diff
@@ -225,6 +231,94 @@ InternalCalculateTSCFromPMTimer (
   return TSCFrequency;
 }
 
+UINT64
+InternalCalculateTSCFromApplePlatformInfo (
+  OUT  UINT64  *FSBFrequency  OPTIONAL
+  )
+{
+  EFI_STATUS                             Status;
+  APPLE_PLATFORM_INFO_DATABASE_PROTOCOL  *PlatformInfo;
+  UINT32                                 Size;
+  VOID                                   *FsbHob;
+  UINT64                                 FsbFreq;
+
+  FsbFreq = 0;
```

Comment:
1. Introduce a guard checking no recursive entrance.
2. Add a Recalculate parameter.
3. Change DEBUG_VERBOSE back to DEBUG_INFO.

Something like that will do:

```
{
STATIC BOOLEAN  ObtainedFreqs = FALSE;
STATIC UINT64 FsbFreq = 0;
STATIC UINT64 TscFreq = 0;

if (Recalculate) {
  ObtainedFreqs = FALSE;
  FsbFreq = TscFreq = 0;
}

if (!ObtainedFreqs) {
  ObtainedFreqs = TRUE;
  /* Reading code */

  TscFreq = InternalConvertAppleFSBToTSCFrequency (FsbFreq);
}

if (FsbFrequency != NULL) {
  *FsbFrequency = FsbFreq;
}

return TscFreq;
}
```